### PR TITLE
Python and Django update utests fixes

### PR DESCRIPTION
**Pull request by TomaszKolek**
…g method not str.

### CODE REVIEW

File: budgetportal/views.py [hunk 1]
```diff
@@ -173,11 +173,11 @@ def department_list_csv(request, financial_year_id, spheres=["national", "provin
             for department in government.departments.all():
                 writer.writerow(
                     {
-                        "government": government.name.encode("utf-8"),
-                        "department_name": department.name.encode("utf-8"),
+                        "government": str(government.name.encode("utf-8")),
```
Comment (jbothma) [inline budgetportal/views.py]: Would it make sense to remove the `.encode` (and then the `str()`) here, since it's already str coming out of the DB?

Comment (TomaszKolek) [reply budgetportal/views.py]: yeah, thats what I'm just doing after a bit of digging into the code. It seems that normal strings are enough for this project.

Thank you! @jbothma

**Comment by jbothma** [comment]
I'm having trouble connecting to UpWork right now - you can email me at jd@openup.org.za if needed

**Comment by TomaszKolek** [comment]
@jbothma going for training now.
2 tests should be fixed, going to investigate the rest right after I'm back.

**Comment by jbothma** [comment]
Nice job!

**Comment by TomaszKolek** [comment]
Getting back

**Comment by jbothma** [comment]
Why do you keep force-pushing?

Don't worry about cleaning commit history - it's fine just to push changes as you fix issues one by one. Force-pushing makes collaboration harder so we prefer just keeping all the commits over time.

**Comment by TomaszKolek** [comment]
Sure @jbothma, most of open source projects I've worked in prefer force pushing in PRs that's why I get used to do it. Probably o keep totally clear history.

Will change that.

**Comment by jbothma** [comment]
Ah cool - that's fine - I guess it makes rebases easier and history
cleaner. I think it's fine to rewrite within a branch but I was worried
that distinct fixes would end up all in one commit which isn't great :)

Today the priority is speed!

**Comment by TomaszKolek** [comment]
2 left! :-)

**Comment by TomaszKolek** [comment]
Actually just one!

**Comment by jbothma** [comment]
Nice! Remember the broken endpoints - see the screenshots and comments in
the original PR.

When they're fixed I'll run your branch and compare with prod and let you
know if I see any remaining issues. I suggest you also do that to get a
sense of what works and what doesn't.

**Comment by jbothma** [comment]
You're doing an awesome job here! Thanks

**Comment by TomaszKolek** [comment]
Current status:
One test left to be fixed.
please replace unicode with str - DONE
No format options available when uploading national infrastructure projects - DONE

**Comment by jbothma** [comment]
Nice! I'll upload some national projects and test them so long

**Comment by jbothma** [comment]
>  No format options available when uploading national infrastructure projects - DONE


Are you sure you've fixed this? 

Screenshot_2020-02-20_15-56-04

**Comment by jbothma** [comment]
Oh I probably need to rebuild to get the new dep

**Comment by TomaszKolek** [comment]
Yeah, its weird, they said in the doc that support is on in 2.0.1 but upgrade help actually.

**Comment by TomaszKolek** [comment]
@jbothma 
http://localhost:8000/2019-20/national/departments/basic-education/

Does it work on production? It seems like it's not related to upgrade.

**Comment by jbothma** [comment]
Yup.

**Comment by TomaszKolek** [comment]
Weird, for me, it looks like some API change. Will be investigating.

**Comment by jbothma** [comment]
I'm quicky setting up a branch based on master to see if I can replicate it

**Comment by jbothma** [comment]
Yup master branch + fresh DB + setup instructions loads basic education
page fine.

By the way, watch out - I've been doing some work with redirecting
.../basic-education to .../basic-education/ which cached a permanent
redirect, but master branch doesn't recognise the URL with trailing slash
so I get 404.

So 404 is probably a caching issue but 500 is definitely upgrade-related.

**Comment by TomaszKolek** [comment]
@jbothma can you check this as well?
budgetportal.tests.test_prov_infra_projects.ProvInfraProjectIRMSnapshotTestCase
For me it fails even on master

**Comment by TomaszKolek** [comment]
confirm budgetportal.tests.test_pages.BasicPagesTestCase 
works on master - will be checking it!

**Comment by TomaszKolek** [comment]
http://localhost:8000/2019-20/national/departments/basic-education/ this works on master indeed - checking as well.

**Comment by TomaszKolek** [comment]
Update about the test:
confirm budgetportal.tests.test_pages.BasicPagesTestCase

I think Django changed a way of evaluating urls in templates, thats why we see it.
Tried to repordude it on master by adding

```
    ds = DepartmentSubprogrammes(department)
    print(ds.get_detail_csv_url())
``` 
just above context variable
to `bugdetportal/views.py`

There is probably something wrong with mocking results mocking.

**Comment by jbothma** [comment]
works for me

```
 (master) $ docker-compose -p datamanager-master run test python manage.py test  budgetportal.tests.test_prov_infra_projects.ProvInfraProjectIRMSnapshotTestCase
```

**Comment by jbothma** [comment]
oh by the way master doesn't have the commands in the setup of each test n that module that clear the index - perhaps there's stuff left after a failed test?

**Comment by TomaszKolek** [comment]
Could you please elaborate on that, not sure if I understand.

I created the PR to check my hypothesis about Django lazy url generting.
https://github.com/vulekamali/datamanager/pull/245
It seems I'm right. Let me know what you think

**Comment by jbothma** [comment]
This is what I was referring to - clearing the index before tests https://github.com/vulekamali/datamanager/pull/239/commits/a089d8d2d3a3d0b94c91faab02e1b6c7ddb7bf99

**Comment by TomaszKolek** [comment]
@jbothma Will check that. Thank you!

http://localhost:8000/2019-20/national/departments/basic-education fixed

**Comment by TomaszKolek** [comment]
http://localhost:8000/json/2019-20/focus.json probably resolved as well

**Comment by TomaszKolek** [comment]
http://localhost:8000/json/2019-20/national/original.json as well

**Comment by TomaszKolek** [comment]
@jbothma lets summarize what is still not done:
1. We still have 2 utests that fail
2. the "department preview pages" - you can reach this by selecting a department in the national or provincial treemap and clicking "epxlore department" (does it work on master? for me it doesn't but maaybe I'm doing something wrong)

**Comment by TomaszKolek** [comment]
I added your commit to my ensuring - still failing.

**Comment by jbothma** [comment]
Merged elsewhere

**Pull request closed without merging**